sdks/go/pkg/beam/core/metrics/metrics.go

```diff
@@ -0,0 +1,479 @@
+// rather than have some marker of uninitialized state, which would otherwise
+// need to be checked for on every update.
+//
+// Metric values are implemented as lightweight proxies of the user provided
+// namespace and name. This allows them to be declared globally, and used in
+// any ParDo. Further, as per the design, they can be declared dynamically
+// at runtime.
+//
+// To handle reporting deltas on the metrics by bundle, metrics
+// are keyed by bundleID,PTransformID,namespace, and name, so metrics that
+// are identical except for bundles are treated as distinct, effectively
+// providing per bundle deltas, since a new value cell is used per bundle.
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/log"
+	"github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+	"github.com/golang/protobuf/ptypes"
+)
+
+// Metric cells are named and scoped by ptransform, and bundle,
+// the latter two of which are only known at runtime. We propagate
+// the PTransformID and BundleID via a context.Context. Consequently
+// using metrics requires the PTransform have a context.Context
+// argument.
+
+type ctxKey string
+
+const bundleKey ctxKey = "beam:bundle"
+const ptransformKey ctxKey = "beam:ptransform"
+
+// SetBundleID sets the id of the current Bundle.
+func SetBundleID(ctx context.Context, id string) context.Context {
+	return context.WithValue(ctx, bundleKey, id)
+}
+
+// SetPTransformID sets the id of the current PTransform.
+func SetPTransformID(ctx context.Context, id string) context.Context {
+	return context.WithValue(ctx, ptransformKey, id)
+}
+
+func getContextKey(ctx context.Context, n name) key {
+	key := key{name: n, bundle: "(bundle id unset)", ptransform: "(ptransform id unset)"}
+	if id := ctx.Value(bundleKey); id != nil {
+		key.bundle = id.(string)
+	}
+	if id := ctx.Value(ptransformKey); id != nil {
+		key.ptransform = id.(string)
+	}
+	return key
+}
+
+// userMetric knows how to convert it's value to a Metrics_User proto.
+type userMetric interface {
+	toProto() *fnexecution_v1.Metrics_User
+}
+
+// name is a pair of strings identifying a specific metric.
+type name struct {
+	namespace, name string
+}
+
+func (n name) String() string {
+	return fmt.Sprintf("%s.%s", n.namespace, n.name)
+}
+
+func newName(ns, n string) name {
+	if len(n) == 0 || len(ns) == 0 {
+		panic(fmt.Sprintf("namespace and name are required to be non-empty, got %q and %q", ns, n))
+	}
+	return name{namespace: ns, name: n}
+}
+
+type key struct {
+	name               name
+	bundle, ptransform string
+}
+
+var (
+	// metricMu protects access to metricStorage
+	metricMu sync.RWMutex
+	// metricStorage is a map of BundleIDs to PtransformIDs to userMetrics.
+	// it permits us to extract metric protos for runners per data Bundle, and
+	// per PTransform.
+	metricStorage = make(map[string]map[string]map[name]userMetric)
```

Comment:
nit: "store"? I think of a package for "storage".

Reply:
Done.